metals/src/main/scala/scala/meta/internal/metals/watcher/SimpleFileWatcher.scala

```diff
@@ -0,0 +1,66 @@
+package scala.meta.internal.metals.watcher
+
+import collection.JavaConverters._
+import java.nio.file.{FileSystems, Path, StandardWatchEventKinds}
+import java.util.concurrent.BlockingQueue
+import scala.concurrent.{ExecutionContext, Future}
+
+object SimpleFileWatcher {
+  /*
+   * returns a function which closes the file watcher when called
+   */
+  def initWatcher(
+    watchFilter: Path => Boolean,
+    watchEventQueue: BlockingQueue[FileWatcherEvent],
+    filesToWatch: Set[Path],
+    directoriesToWatch: Set[Path],
+  )(implicit ec: ExecutionContext): () => Unit = {
+    scribe.error(s"RFERGUSON INIT WATCHER")
+    val watchService = FileSystems.getDefault().newWatchService();
+    filesToWatch.filter(watchFilter).foreach { path =>
+      scribe.info(s"watching parent directory of file $path")
+      // watchservice only watches directories
+      // this technically could watch things that weren't in the build, but shared a directory, but what can you do?
+      path.getParent().register(
+        watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_DELETE,
+        StandardWatchEventKinds.ENTRY_MODIFY,
+        //StandardWatchEventKinds.OVERFLOW,
+      )
+    }
+    directoriesToWatch.filter(watchFilter).foreach { path =>
+      // we don't really have to watch the parent directory, but then we don't have to have special logic for these down below
+      scribe.info(s"watching parent directory of directory $path")
+      path.getParent().register(
+        watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_DELETE,
+        StandardWatchEventKinds.ENTRY_MODIFY,
+      )
+    }
+    Future {
+      while (true) {
+        val key = watchService.take()
+        val directory: Path = key.watchable().asInstanceOf[Path]
+        if (key != null) {
+          key.pollEvents().asScala.foreach { event =>
+            val pathFromEventCtx = event.context().asInstanceOf[Path]
+            val path = directory.resolve(pathFromEventCtx)
+            if (watchFilter(path)) {
+              scribe.info(s"got ${event.kind()} event for path: ${path}")
+              event.kind() match {
+                case StandardWatchEventKinds.ENTRY_CREATE | StandardWatchEventKinds.ENTRY_MODIFY =>
+                  watchEventQueue.add(FileWatcherEvent.createOrModify(path))
+                case StandardWatchEventKinds.ENTRY_DELETE =>
+                  watchEventQueue.add(FileWatcherEvent.delete(path))
+              }
+            }
+          }
+          key.reset()
+        }
+      }
+    }
+    watchService.close
```

Comment:
When the watch service is closed, will the `Future` spawned above stop?

I want to ensure there won't be any memory leaks.

Reply:
Yeah, if there's a take() call waiting or if you call take() after it's already closed, it'll throw an exception, so it should clean itself up right away.